File: nature-platform-web/src/auth-service.ts
/**
 * @input apiClient and ApiResponse from local HTTP layer
 * @output login(), fetchCurrentUser(), fetchDingTalkLoginUrl() auth service functions with role/resource/menu profile contracts
 * @position Frontend application service layer for authentication and RBAC bootstrap API calls
 * @doc-sync Update this header and folder INDEX.md when this file changes.
 */
import { apiClient, type ApiResponse } from "./api";

export interface LoginResponse {
  token: string;
  username: string;
  mustChangePassword: boolean;
}

export interface MenuTreeNode {
  resourceKey: string;
  resourceName: string;
  resourceType: "GROUP" | "PAGE";
  parentKey?: string;
  routePath?: string;
  icon?: string;
  sortOrder: number;
  enabled: boolean;
  builtIn: boolean;
  description?: string;
  children: MenuTreeNode[];
}

export interface CurrentUser {
  username: string;
  displayName: string;
  roles: string[];
  resources: string[];
  permissions: string[];
  menuTree: MenuTreeNode[];
  timezone: string;
}

export async function login(username: string, password: string): Promise<LoginResponse> {
  const response = await apiClient.post<ApiResponse<LoginResponse>>("/auth/login", {
    username,
    password
  });
  return response.data.data;
}

export async function fetchCurrentUser(): Promise<CurrentUser> {
  const response = await apiClient.get<ApiResponse<CurrentUser>>("/auth/me");
  return response.data.data;
}

export async function fetchDingTalkLoginUrl(): Promise<string> {
  const response = await apiClient.get<ApiResponse<{ url: string }>>("/auth/dingtalk/url");
  return response.data.data.url;
}

File: nature-platform-api/src/main/java/com/nature/platform/AuthService.java
/**
 * @input JwtTokenService, UserAccountService, AdminAccessService, and AdminResourceService for IAM-aware profile aggregation
 * @output login() and currentUserProfile() methods for local authentication and role/resource/menu bootstrap payloads
 * @position Authentication application service layer bridging account/role persistence to token session and frontend RBAC context
 * @doc-sync Update this header and folder INDEX.md when this file changes.
 */
package com.nature.platform;

import java.util.List;
import java.util.Map;
import org.springframework.stereotype.Service;

@Service
public class AuthService {
  private final JwtTokenService jwtTokenService;
  private final UserAccountService userAccountService;
  private final AdminAccessService adminAccessService;
  private final AdminResourceService adminResourceService;
  private final DingTalkAuthService dingTalkAuthService;

  public AuthService(
      JwtTokenService jwtTokenService,
      UserAccountService userAccountService,
      AdminAccessService adminAccessService,
      AdminResourceService adminResourceService,
      DingTalkAuthService dingTalkAuthService) {
    this.jwtTokenService = jwtTokenService;
    this.userAccountService = userAccountService;
    this.adminAccessService = adminAccessService;
    this.adminResourceService = adminResourceService;
    this.dingTalkAuthService = dingTalkAuthService;
  }

  public LoginResponse login(LoginRequest request) {
    UserAccountService.UserAccount userAccount =
        userAccountService
            .findByUsername(request.getUsername())
            .orElseThrow(() -> new IllegalArgumentException("invalid username or password"));
    if (!userAccount.enabled() || !userAccount.passwordHash().equals(request.getPassword())) {
      throw new IllegalArgumentException("invalid username or password");
    }

    List<String> scopes = userAccountService.listRoles(userAccount.username());
    String token = jwtTokenService.generateToken(request.getUsername(), scopes);
    return new LoginResponse(token, request.getUsername(), userAccount.mustChangePassword());
  }

  public LoginResponse loginByDingTalk(String authCode) {
    return dingTalkAuthService.loginByAuthCode(authCode);
  }

  public void changePassword(String username, String newPassword) {
    userAccountService.updatePassword(username, newPassword);
  }

  public Map<String, Object> currentUserProfile(String username) {
    UserAccountService.UserAccount userAccount =
        userAccountService
            .findByUsername(username)
            .orElse(
                new UserAccountService.UserAccount(
                    0L, username, "", username, true, null, false, "LOCAL", null, null));
    List<String> roles = userAccountService.listRoles(username);
    List<String> resources = adminAccessService.listResources(username);
    List<AdminResourceRecord> menuTree = adminResourceService.listResourceTreeForUser(username);
    return Map.of(
        "username", username,
        "displayName", userAccount.displayName(),
        "mustChangePassword", userAccount.mustChangePassword(),
        "roles", roles,
        "permissions", resources,
        "resources", resources,
        "menuTree", menuTree,
        "timezone", "Asia/Shanghai");
  }
}
